### PR TITLE
Optimise conversion of geometry from OGR -> QGIS 

### PR DESCRIPTION
**Pull request by nyalldawson**
Avoid conversion to/from WKB at OGR/QGIS side, and just directly utilise OGR geometry API to construct QGIS geometries. Implemented for point/linestrings only for now.

Shaves ~10% off rendering time for a large point layer (GPKG).

### CODE REVIEW

**Comment by nyalldawson** [comment]
@rouault what do you think? Good idea? Bad idea? Any possible side effects?

File: src/core/qgsogrutils.cpp [hunk 1]
```diff
@@ -274,6 +275,55 @@ QgsGeometry QgsOgrUtils::ogrGeometryToQgsGeometry( OGRGeometryH geom )
...
+      QVector< double > m;
+      if ( QgsWkbTypes::hasM( wkbType ) )
+      {
+        m.resize( count );
+        pm = m.data();
+      }
+      double *px = x.data();
+      double *py = y.data();
+      double zz, mm;
+      for ( int i = 0; i < count; ++i )
```
Comment (rouault) [inline src/core/qgsogrutils.cpp]: Using OGR_G_GetPointsZM() to directly fill the arrays should probably be faster

**Comment by wonder-sk** [comment]
Yeah I was always wondering why are we using WKB to pass geometry data instead of OGR API... +1 to use OGR API if it is faster.

File: src/core/qgsogrutils.cpp (hunk 1 above)
Comment (nyalldawson) [reply src/core/qgsogrutils.cpp]: Thanks -- updated.

**Comment by nyalldawson** [comment]
@wonder-sk 
> Yeah I was always wondering why are we using WKB to pass geometry data instead of OGR API... +1 to use OGR API if it is faster.

I suspect this predates the new geometry engine from 2.10 -- back from the olden days when QgsGeometry was just a wkb store internally. But it's confused me too!

**Comment by luipir** [comment]
+1 to reduce redundant classes

**Comment by NathanW2** [comment]
Nice work.

File: src/core/qgsogrutils.cpp (hunk 1 above)
Comment (nyalldawson) [reply src/core/qgsogrutils.cpp]: @rouault Are the docs for OGR_G_GetPointsZM incorrect? They state:

    pabyX a buffer of at least (sizeof(double) * nXStride * nPointCount) bytes, may be NULL.

Shouldn't this be just:

    pabyX a buffer of at least (nXStride * nPointCount) bytes, may be NULL.

Comment (rouault) [reply src/core/qgsogrutils.cpp]: @nyalldawson You're right. The docs were incorrect. I've just fixed then

Comment (nyalldawson) [reply src/core/qgsogrutils.cpp]: Thanks!

**Comment by wonder-sk** [comment]
Looking at the final code... what is the actual difference between `LineString25D` and `LineStringZ`? Aren't those `25D` types some kind of legacy and couldn't we simply auto-convert between `25D` and `Z` WKB types when working with OGR library? I am a bit worried that there may be plenty of code in QGIS that can't handle the `25D` variant correctly...

I am wondering whether we can avoid `bool is25DType` parameter in `QgsLineString` constructor... or if it is needed, shouldn't other constructor variants use that as well?

**Comment by nyalldawson** [comment]
@wonder-sk 

I've tried to exactly emulate the current behaviour here -- I didn't want to make a possibly dangerous change here. We could/should(?) drop the 25d I think, but we'd need to proceed carefully...

**Pull request merged**